core/templates/pages/exploration-editor-page/services/changes-in-human-readable-form.service.ts

```diff
@@ -39,8 +39,8 @@ export class ChangesInHumanReadableFormService {
   makeRulesListHumanReadable(answerGroupValue: any): Array<HTMLElement> {
     let rulesList = [];
     answerGroupValue.rules.forEach((rule) => {
-      let ruleElm = this.document.createElement('<li></li>');
-      ruleElm.html('<p>Type: ' + rule.type + '</p>');
+      let ruleElm = this.document.createElement('li');
+      ruleElm.innerHTML = '<p>Type: ' + rule.type + '</p>';
```

Comment:
Please don't use innerHtml, it has security problems. Find another way (perhaps similar to the previous way, or do something declarative with Angular templates).

Reply:
(Probably best to use an angular template as we do for everything else, and only compile the data in JS -- though I realize that's a bigger change than might have been originally intended.)

Reply:
@seanlip I couldn't figure out how to use Angular templates for services. Also, I don't see any other services using templates, am I missing something?
BTW, I found out that `textContent` is a safer alternative to `innerHTML` since it gets and injects only text content, not markup, and thereby prevents cross-site scripting attacks. ([Reference](https://gomakethings.com/preventing-cross-site-scripting-attacks-when-using-innerhtml-in-vanilla-javascript/)) Should I make changes and use `textContent` instead?

Reply:
Ah -- let me be clearer. This service is currently html-izing the data and passing it to the modal. Instead of doing that, it is better to create some kind of domain object to house the data, store only the raw/compiled (non-htmlized) data in that domain object, and do all the templating/html stuff in the directive, using the standard way for doing this (having a template etc.). Does that clarify?

Also, textContent is indeed safer, but I don't think you can construct html using it, so I think the above might be a better way to go.

Reply:
I can do that but I'm a bit confused, do you have any article reference about this approach? I'd to understand it further.

Reply:
There isn't an article, sorry -- but there are many examples already, i.e. basically almost every existing directive/controller in the codebase.

Pretend you're building this modal from scratch. The usual way to do it would be to make a modal template in a template file, and then assemble the data in JS, and use that data to fill the template. That is the way this should be done, instead of manipulating HTML in the JS -- the JS should be defining and populating variables like $ctrl.XXX and these can be used in the template to interpolate. When you are done, the final result should have no paragraph (and other) HTML tags, or any fragments of HTML strings, in the JS at all.

Does that make sense? Please let me know if I can clarify further. Thanks!

Reply:
Hi @seanlip one of my comments about `textContent` were pending. I think I've got what you mean, but I don't know how to do it in this case. I just refactored the service because it was calling wrongly some `document` methods, I've followed the logic it was before.

Reply:
Oh, sorry, somehow I missed that other comment.

To that question, I don't know which the best approach is, but textContent is safer than innerHTML. However you can't put html in textContent -- it's just for the raw text within HTML elements.